File: Week_02/Practicum/Maximum Ice Cream Bars.cpp
class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        int count = 0;

        sort(costs.begin(), costs.end());

        for (int i = 0; i < costs.size(); i++) {
            coins -= costs[i];
            if (coins < 0)
                break;
            count++;
        }
        return count;
    }
};


class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        const long max = (int)10e5 + 1;
        long sortedCosts[max] = {0};

        for(long i = 0; i < costs.size(); i++) {
            sortedCosts[costs[i]] ++;
        }

        long count = 0;
        for(long i = 0; i < max; i++) {
            
            if(i>coins){
                break;
            }

            long br = sortedCosts[i];

            if(br*i > coins){
                count += coins/i;

                break;
            }
            
            coins -= br*i;
            count += br;
        }

        return count;
    }
};

File: Week_02/Practicum/python/Maximum Ice Cream Bars.py
class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        costs.sort()
        
        total_ice_creams = 0
        i = 0
        while i < len(costs) and costs[i] <= coins:
            total_ice_creams += 1
            coins -= costs[i]
            i += 1

        return total_ice_creams
            

        